diff --git a/deps/openssl/config/Makefile b/deps/openssl/config/Makefile
index 5ed49e4d..b40ae2bb 100644
--- a/deps/openssl/config/Makefile
+++ b/deps/openssl/config/Makefile
@@ -9,11 +9,11 @@ endif
 PERL    = perl
 
 # Supported architecture list
-ASM_ARCHS = aix64-gcc-as BSD-x86 BSD-x86_64 \
-darwin64-x86_64-cc darwin-i386-cc darwin64-arm64-cc linux-aarch64 \
-linux-armv4 linux-elf linux-x86_64 \
-linux-ppc64le linux32-s390x linux64-s390x linux64-mips64\
-solaris-x86-gcc solaris64-x86_64-gcc VC-WIN64A VC-WIN32
+ASM_ARCHS = aix64-gcc-as BSD-x86 BSD-x86_64 BSD-aarch64 darwin64-x86_64-cc \
+darwin-i386-cc darwin64-arm64-cc darwin-armv7-cc linux-aarch64 linux-armv4 \
+linux-x86 linux-x86_64 linux-ppc linux-ppc64le linux32-s390x linux64-s390x \
+linux64-mips64 linux-mips32 qnx-x86 qnx-arm solaris-x86-gcc \
+solaris64-x86_64-gcc VC-WIN64A VC-WIN32
 
 NO_ASM_ARCHS = VC-WIN64-ARM linux64-riscv64
 
@@ -24,8 +24,9 @@ CONFIGURE = ./Configure
 # no-comp: against CRIME attack
 # no-shared: openssl-cli needs static link
 # no-afalgeng: old Linux kernel < 4.0 does not support it
+# no-async: maximum portability
 # enable-ssl-trace: cause the optional SSL_trace API to be built
-COPTS =  no-comp no-shared no-afalgeng enable-ssl-trace enable-fips
+COPTS =  no-comp no-shared no-afalgeng no-async enable-ssl-trace enable-fips
 
 # disable platform check in Configure
 NO_WARN_ENV = CONFIGURE_CHECKER_WARN=1
diff --git a/deps/openssl/config/generate_gypi.pl b/deps/openssl/config/generate_gypi.pl
index be7d357f..bd5ad24d 100755
--- a/deps/openssl/config/generate_gypi.pl
+++ b/deps/openssl/config/generate_gypi.pl
@@ -361,9 +361,37 @@ open(CLGYPI, "> ./archs/$arch/$asm/openssl-cl.gypi");
 print CLGYPI "$clgypi";
 close(CLGYPI);
 
+# Create meson.build
+my $mtemplate =
+    Text::Template->new(TYPE => 'FILE',
+                        SOURCE => 'meson.build.tmpl',
+                        DELIMITERS => [ "%%-", "-%%" ]
+                        );
+
+my $meson = $mtemplate->fill_in(
+    HASH => {
+        libssl_srcs => \@libssl_srcs,
+        libcrypto_srcs => \@libcrypto_srcs,
+        lib_defines => \@lib_defines,
+        generated_srcs => \@generated_srcs,
+        apps_openssl_srcs => \@apps_openssl_srcs,
+        libapps_srcs => \@libapps_srcs,
+        config => \%config,
+        target => \%target,
+        cflags => \@cflags,
+        asm => \$asm,
+        arch => \$arch,
+        lib_cppflags => \@lib_cppflags,
+        is_win => \$is_win,
+    });
+
+open(MESON, "> ./archs/$arch/$asm/meson.build");
+print MESON "$meson";
+close(MESON);
+
 # Clean Up
 my $cmd2 ="cd $src_dir; make -f $makefile clean; make -f $makefile distclean;" .
-    "git clean -f $src_dir/crypto";
+    "git clean -f crypto";
 system($cmd2) == 0 or die "Error in system($cmd2)";
 
 
